### ISSUE TITLE
Train set has different image size from the eval set.  How to train and eval on different image sizes in the same training loop?

### ISSUE THREAD

**Issue by holladayinnn** [ultralytics/ultralytics#14607]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

What would be the easiest way to train a model, but after every epoch, eval on a dataset with a different image size from the train set?  The config only allows specifying  a single integer for the imgsz, but train and eval have different image sizes.  Thanks!

**Comment by pderrenger** [comment]
@holladayinnn to train and evaluate on different image sizes within the same training loop, you can modify the training script to adjust the image size dynamically before each evaluation step. While the configuration file allows specifying a single image size, you can override this setting programmatically. Ensure you are using the latest version of the Ultralytics package to avoid any potential issues. If you encounter any problems, please let us know.

**Comment by glenn-jocher** [comment]
@holladayinnn you can train no matter what image sizes are in your train or val sets, they are all automatically resized by `ultralytics` to `imgsz` before being used, no action is required on your part. 

If you'd like to update the image size used you can customize easily with i.e. `imgsz=1280` or `imgsz=320` at train/val/predict/export time.

**Comment by Y-T-G** [comment]
You can use a callback hook to add custom functionality.

You could change model.overrides["imgsz"] `on_val_start` and change it again back to default on `on_val_end`.

**Comment by Y-T-G** [comment]
https://github.com/ultralytics/ultralytics/blob/205d560ebc73a2799c598ba746e348ed08e6cdea/ultralytics/utils/callbacks/hub.py#L83

**Comment by holladayinnn** [comment]
Is something like this what you have in mind?  Am I missing anything here?

`def resize_val_image(validator):
    config_file = '/home/ubuntu/.local/lib/python3.8/site-packages/ultralytics/cfg/default.yaml'
    with open(config_file) as f:
      cfg = yaml.safe_load(f)
    validator.dataloader.dataset.imgsz = check_imgsz(val_image_size)
    transforms = Compose([LetterBox(new_shape=(validator.dataloader.dataset.imgsz, validator.dataloader.dataset.imgsz), scaleup=False)])
    t = Format(
                bbox_format="xywh",
                normalize=True,
                return_mask=validator.dataloader.dataset.use_segments,
                return_keypoint=validator.dataloader.dataset.use_keypoints,
                return_obb=validator.dataloader.dataset.use_obb,
                batch_idx=True,
                mask_ratio=cfg['mask_ratio'],
                mask_overlap=cfg['overlap_mask'],
                bgr=cfg['bgr'] if validator.dataloader.dataset.augment else 0.0,  # only affect training.
            )
    transforms.append(t)
    validator.dataloader.dataset.transforms = transforms

model.add_callback('on_val_start', resize_val_image)
`

**Comment by holladayinnn** [comment]
Looks like its still resizing :(

**Comment by pderrenger** [comment]
It seems like the image size is still being resized to the training set size. To address this, ensure that the callback correctly modifies the validation image size before the evaluation step. Double-check that the callback is properly registered and that the `imgsz` is being set correctly within the validation dataset. If the issue persists, please verify that you are using the latest version of the Ultralytics package.

**Comment by holladayinnn** [comment]
I think I found a solution that works for me.

```
class DetectionTrainer2(DetectionTrainer):
    def build_dataset(self, img_path, mode="train", batch=None):
        """
        Build YOLO Dataset.

        Args:
            img_path (str): Path to the folder containing images.
            mode (str): `train` mode or `val` mode, users are able to customize different augmentations for each mode.
            batch (int, optional): Size of batches, this is for `rect`. Defaults to None.
        """
        if mode == 'train':
            args = self.args
        else:
            args = copy(self.args)
            args.imgsz = val_img_size
        gs = max(int(de_parallel(self.model).stride.max() if self.model else 0), 32)
        return build_yolo_dataset(args, img_path, batch, self.data, mode=mode, rect=mode == "val", stride=gs)

    def get_validator(self):
        """Returns a DetectionValidator for YOLO model validation."""
        self.loss_names = "box_loss", "cls_loss", "dfl_loss"
        args = copy(self.args)
        args.imgsz = val_img_size
        return yolo.detect.DetectionValidator(self.test_loader, save_dir=self.save_dir, args=args, _callbacks=self.callbacks)

config_file = '/home/ubuntu/.local/lib/python3.8/site-packages/ultralytics/cfg/default.yaml'
with open(config_file) as f:
    cfg = yaml.safe_load(f)
model = YOLO(weights)
trainer = DetectionTrainer2(overrides=cfg)
trainer.model = trainer.get_model(model.model.yaml, model.model, verbose=False)
trainer.train()
```

Only issue now is the batch size for the validation step is too large and would like to individually change that too without having to reduce the train batch size.

**Comment by pderrenger** [comment]
Thank you for sharing your solution. To address the issue of setting different batch sizes for training and validation, you can modify the `DetectionTrainer2` class to accept separate batch size parameters for each mode. This way, you can customize the batch size independently for training and validation without affecting each other. If you encounter any issues, please ensure you are using the latest version of the Ultralytics package. If you need further assistance, feel free to ask!

**Comment by holladayinnn** [comment]
Got it working.  Here is the class:
```
class DetectionTrainer2(DetectionTrainer):
    def __init__(self, val_imgsz, val_batch_size, cfg=DEFAULT_CFG, overrides=None, _callbacks=None):
        self.val_imgsz = val_imgsz
        self.val_batch_size = val_batch_size
        super().__init__(cfg, overrides, _callbacks)

    def build_dataset(self, img_path, mode="train", batch=None):
        """
        Build YOLO Dataset.

        Args:
            img_path (str): Path to the folder containing images.
            mode (str): `train` mode or `val` mode, users are able to customize different augmentations for each mode.
            batch (int, optional): Size of batches, this is for `rect`. Defaults to None.
        """
        if mode == 'train':
            args = self.args
        else:
            args = copy(self.args)
            args.imgsz = self.val_imgsz
            args.batch = self.val_batch_size
        gs = max(int(de_parallel(self.model).stride.max() if self.model else 0), 32)
        return build_yolo_dataset(args, img_path, batch, self.data, mode=mode, rect=mode == "val", stride=gs) 

    def get_validator(self):
        """Returns a DetectionValidator for YOLO model validation."""
        self.loss_names = "box_loss", "cls_loss", "dfl_loss"
        args = copy(self.args)
        args.imgsz = self.val_imgsz
        args.batch = self.val_batch_size
        return yolo.detect.DetectionValidator(self.test_loader, save_dir=self.save_dir, args=args, _callbacks=self.callbacks)

    def get_dataloader(self, dataset_path, batch_size=16, rank=0, mode="train"):
        """Construct and return dataloader."""
        if mode != "train":
            batch_size = self.val_batch_size
        assert mode in {"train", "val"}, f"Mode must be 'train' or 'val', not {mode}."
        with torch_distributed_zero_first(rank):  # init dataset *.cache only once if DDP
            dataset = self.build_dataset(dataset_path, mode, batch_size)
        shuffle = mode == "train"
        if getattr(dataset, "rect", False) and shuffle:
            LOGGER.warning("WARNING ⚠️ 'rect=True' is incompatible with DataLoader shuffle, setting shuffle=False")
            shuffle = False
        workers = self.args.workers if mode == "train" else self.args.workers * 2
        return build_dataloader(dataset, batch_size, workers, shuffle, rank)  # return dataloader
```

**Comment by pderrenger** [comment]
@holladayinnn great to hear you got it working! Your approach to customizing the `DetectionTrainer2` class for different image sizes and batch sizes during training and validation looks solid. If you encounter any further issues or have additional questions, please ensure you are using the latest version of the Ultralytics package for the best compatibility and performance. Feel free to reach out here for more assistance.
